src/schema.ts

```diff
@@ -173,7 +181,281 @@ class SchemaDefinition implements SchemaDefinitionType {
+        const parents: string[] = [];
+
+        for (const parent of model.inherits) {
+            parents.push(parent);
+            parents.push(...this.ancestors(parent));
+        }
+        return parents;
+    }
+
+    children(modelName: string): string[] {
+        const model = this.get(modelName);
+        return this.subclassMapping[model.name] || [];
+    }
+
+    /**
+     * Get a list of models (including the current model) for all descendants of the current model
+     *
+     * @param excludeAbstract exclude abstract models
+     *
+     * @returns the array of descendant models
+     */
+    descendants(
+        modelName: string,
+        opt: { excludeAbstract?: boolean; includeSelf?: boolean } = {},
+    ): string[] {
+        const { excludeAbstract = false, includeSelf = false } = opt;
+        const descendants: string[] = includeSelf
+            ? [modelName]
+            : [];
+        const queue: string[] = this.children(modelName).slice();
+
+        while (queue.length > 0) {
+            const child = queue.shift() as typeof queue[number];
+
+            if (descendants.includes(child)) {
+                continue;
+            }
+            descendants.push(child);
+            queue.push(...this.children(child));
+        }
+        return descendants.filter((child) => {
+            const descendantModel = this.get(child, true);
+            return (!excludeAbstract || !descendantModel.isAbstract);
+        });
+    }
+
+    /**
+     * Returns a set of properties from this class and all subclasses
+     * @type {Array.<PropertyDefinition>}
+     */
+    queryableProperties(modelName: string) {
+        const model = this.get(modelName);
+        const queue = Array.from(this.children(modelName));
+        const queryProps = this.getProperties(model.name);
+
+        while (queue.length > 0) {
+            const curr = queue.shift() as typeof queue[number];
+            const currModel = this.get(curr);
+
+            for (const prop of Object.values(this.getProperties(currModel.name))) {
+                if (queryProps[prop.name] === undefined) { // first model to declare is used
+                    queryProps[prop.name] = prop;
+                }
+            }
+            queue.push(...this.children(curr));
+        }
+        return queryProps;
+    }
+
+    /**
+     * Check if this model inherits a property from a parent model
+     */
+    inheritsProperty(modelName: string, propName: string) {
+        const model = this.get(modelName);
+
+        for (const parent of model.inherits) {
+            const parentModel = this.get(parent);
+
+            if (parentModel.properties[propName] !== undefined) {
+                return true;
+            }
+            if (this.inheritsProperty(parent, propName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks a single record to ensure it matches the expected pattern for this class model
```

Comment:
this docstring makes it sounds like it only validates the record. there should be something describing that it returns a copy of the record with modifications applied based upon what settings were provided

Reply:
9b6ae2e